File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/sync/library/items/specs/FakeDeferredStoredItemAccess.java
package com.lasthopesoftware.bluewater.client.sync.library.items.specs;

import com.lasthopesoftware.bluewater.client.library.items.IItem;
import com.lasthopesoftware.bluewater.client.sync.library.items.IStoredItemAccess;
import com.lasthopesoftware.bluewater.client.sync.library.items.StoredItem;
import com.namehillsoftware.handoff.Messenger;
import com.namehillsoftware.handoff.promises.Promise;

import java.util.Collection;

public abstract class FakeDeferredStoredItemAccess implements IStoredItemAccess {

	private Messenger<Collection<StoredItem>> messenger;

	public void resolveStoredItems() {
		if (messenger != null) {
			messenger.sendResolution(getStoredItems());
		}
	}

	protected abstract Collection<StoredItem> getStoredItems();

	@Override
	public void toggleSync(IItem item, boolean enable) {

	}

	@Override
	public Promise<Boolean> isItemMarkedForSync(IItem item) {
		return new Promise<>(false);
	}

	@Override
	public Promise<Collection<StoredItem>> promiseStoredItems() {
		return new Promise<>((m -> messenger = m));
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/FakeDeferredStoredItemAccess.kt
package com.lasthopesoftware.bluewater.client.stored.library.items

import com.lasthopesoftware.bluewater.client.browsing.items.IItem
import com.lasthopesoftware.bluewater.client.browsing.items.KeyedIdentifier
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.shared.promises.extensions.toPromise
import com.namehillsoftware.handoff.Messenger
import com.namehillsoftware.handoff.promises.Promise

abstract class FakeDeferredStoredItemAccess : AccessStoredItems {

	private var messenger: Messenger<Collection<StoredItem>>? = null

	fun resolveStoredItems() {
		messenger?.sendResolution(storedItems)
	}

	protected abstract val storedItems: Collection<StoredItem>

	override fun disableAllLibraryItems(libraryId: LibraryId): Promise<Unit> = Unit.toPromise()

	override fun toggleSync(libraryId: LibraryId, itemId: KeyedIdentifier) = true.toPromise()
	override fun toggleSync(libraryId: LibraryId, item: IItem, enable: Boolean) = Unit.toPromise()
	override fun toggleSync(libraryId: LibraryId, itemId: KeyedIdentifier, enable: Boolean) = Unit.toPromise()
	override fun isItemMarkedForSync(libraryId: LibraryId, item: IItem) = false.toPromise()

	override fun isItemMarkedForSync(libraryId: LibraryId, itemId: KeyedIdentifier) = false.toPromise()

	override fun promiseStoredItems(libraryId: LibraryId): Promise<Collection<StoredItem>> {
		return Promise { m -> messenger = m }
	}
}
